### PR TITLE
meeting scheduling algorithm and tests for calendar project

### PR DESCRIPTION
**Pull request by emaohn**

### CODE REVIEW

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 1]
```diff
@@ -22,6 +22,13 @@
...
    */
   public String greet(String name) {
-    return "Hello " + name;
+    // String result = name;
+    // // find start of name
+    // for (int i = 0; i < name.length; i++) {
+    //   if (name.charAt(i) != ' ') {
+    //     result = name.substring(i);
+    //   }
+    // }
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java]: For cleanliness, better to delete this than comment it out. If you might want it later you can put it on another branch instead.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,118 @@
...
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return mandatoryTimeRanges;
+    }
+    // initialize time range with whole day
+    mandatoryTimeRanges.add(TimeRange.WHOLE_DAY);
+    optionalTimeRanges.add(TimeRange.WHOLE_DAY);
+    Collection<String> MandatoryAttendants = request.getAttendees();
+    Collection<String> optionalAttendants = request.getOptionalAttendees();
+    for(Event e: events) {
+      boolean checkMandatoryAttendee = MandatoryAttendants.size() == 0 ? false : true;
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This variable confuses me a bit. Isn't this redundant because if `MandatoryAttendants` is empty, it will never contain the attendee? Why do you override it lower down when it will just be changed by the next iteration?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,118 @@
...
+            mandatoryIndex++;
+        }
+    }
+    if (MandatoryAttendants.size() != 0 && overlapTimeRanges.size() == 0) {
+        return mandatoryTimeRanges;
+    }    
+    return overlapTimeRanges;
   }
+
+    private TimeRange getOverlap(TimeRange mandatoryRange, TimeRange optionalRange) {
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: This is nicely laid out.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,118 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*; 
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: It took me a bit to figure out what events meant in this context, using a variable name like existingEvents or previousEvents might help make the code a little more clear.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,118 @@
...
+      if (optionalAttendants.size() != 0 && numOptionalUnavailable == optionalAttendants.size()) {
+          handleEventConflict(optionalTimeRanges, e, request);
+      }
+    }
+
+    // See if there are any valid overlaps between the mandatory time ranges and optional time ranges
+    ArrayList<TimeRange> overlapTimeRanges = new ArrayList<TimeRange>();
+    int mandatoryIndex = 0;
+    int optionalIndex = 0;
+    while (mandatoryIndex < mandatoryTimeRanges.size() && optionalIndex < optionalTimeRanges.size()) {
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: This loop seems to rely on the time ranges being sorted, but I don't see a sort or anything else that guarantees order in here (though could be I'm missing it). Make sure that your assumptions are guaranteed or you can get weird edge case behavior.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,118 @@
...
+                // case 1: current timerange is within event timerange => no part of timerange works so remove
+                if (eventRange.contains(curRange)) {
+                    timeranges.remove(i);
+                // case 2: event timerange is within current timerange
+                } else if (curRange.contains(eventRange)) {
+                    int numNewRanges = 0;
+                    if (eventRange.start() - curRange.start() >= request.getDuration()) {
+                        timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(curRange.start(), eventRange.start(), false));                    }
+                    if (curRange.end() - eventRange.end() >= request.getDuration()) {
+                    timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(eventRange.end(), curRange.end(), false));
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: Please indent

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,118 @@
...
+        }
+        if (mandatoryRange.start() < optionalRange.start()) {
+            return TimeRange.fromStartEnd(optionalRange.start(), mandatoryRange.end(), true);
+        } else {
+            return TimeRange.fromStartEnd(mandatoryRange.start(), optionalRange.end(), true);
+        }
+    }
+
+    private void handleEventConflict(ArrayList<TimeRange> timeranges, Event e, MeetingRequest request) {
+        for (int i = 0; i < timeranges.size();) {
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: So what you're doing here is iterating over indices and deleting items. I believe this is going to be problematic because as you remove items, the list gets shorter and the indices mean different things since ArrayLists are dynamic.

As an example, I think this snippet:
```
ArrayList<String> x = new ArrayList<String>(Array.asList("hi", "bye"));
for (int i = 0; i < x.size();) {
  System.out.println(i);
  System.out.println(x[i]);
  x.remove(i);
  System.out.println(x[i]);
}
```

Would output:
```
0
hi
bye
```

Definitely double check this belief since Java is not my best language. Here's a discussion I used as a source that might be helpful if you need to do something else: https://stackoverflow.com/questions/10431981/remove-elements-from-collection-while-iterating

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 8]
```diff
@@ -270,5 +272,111 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void optionalAttendeesAreConsidered() {
```
Comment (cthamrun) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:300]: Nit: consider adding a test scenario to reflect/test the behavior where there is a time slot that works for person A and person B but none that works for person C (I expect that we ignore person C since that person is optional and just schedule as per usual between person A and B).

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 9]
```diff
@@ -270,5 +272,111 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void everyMandatoryAttendeesIsPrioritized() {
```
Comment (ganewman42) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:277]: Nice test coverage.

**Review by ganewman42 — COMMENTED** [review]
A few nits and what might be a few subtle bugs. Take a look and let me know if you have any questions.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (emaohn) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: That makes sense--this was just the method header already given to us in the starter code but I agree and I was confused as well when I started.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (emaohn) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: My algorithm for modifying the timeranges list should ensure that they're being split and inserted in an ordered manner. If you go down a bit you'll see the insertion part when I add to the timeranges. Maybe there is a bug that I don't know of?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (emaohn) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes you're correct! I believe I handled this situation however..which is why I don't automatically indent i (i++ in the for loop header) and instead manually increment i as necessary. Perhaps this isn't best code practice?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,118 @@
...
+    // if meeting duration is longer than a day, it is not possible to hold meeting
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return mandatoryTimeRanges;
+    }
+    // initialize time range with whole day
+    mandatoryTimeRanges.add(TimeRange.WHOLE_DAY);
+    optionalTimeRanges.add(TimeRange.WHOLE_DAY);
+    Collection<String> MandatoryAttendants = request.getAttendees();
+    Collection<String> optionalAttendants = request.getOptionalAttendees();
+    for(Event e: events) {
```
Comment (cthamrun) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: space after "for" and space after "e" (i.e. "for (Event e : events) {") for consistency within the file.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,118 @@
...
+      boolean checkMandatoryAttendee = MandatoryAttendants.size() == 0 ? false : true;
+      int numOptionalUnavailable = 0;
+      for(String attendee: e.getAttendees()) {
```
Comment (cthamrun) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: same here as comment above.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,118 @@
...
+                if (eventRange.contains(curRange)) {
+                    timeranges.remove(i);
+                // case 2: event timerange is within current timerange
+                } else if (curRange.contains(eventRange)) {
+                    int numNewRanges = 0;
+                    if (eventRange.start() - curRange.start() >= request.getDuration()) {
+                        timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(curRange.start(), eventRange.start(), false));                    }
+                    if (curRange.end() - eventRange.end() >= request.getDuration()) {
+                    timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(eventRange.end(), curRange.end(), false));
+                    }
```
Comment (cthamrun) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:113]: Nit: make sure indentation and curly braces placement is consistent in this chunk here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -14,10 +14,118 @@
...
+            TimeRange eventRange = e.getWhen();
+            if (curRange.overlaps(eventRange)) {
+                // case 1: current timerange is within event timerange => no part of timerange works so remove
+                if (eventRange.contains(curRange)) {
+                    timeranges.remove(i);
+                // case 2: event timerange is within current timerange
+                } else if (curRange.contains(eventRange)) {
+                    int numNewRanges = 0;
+                    if (eventRange.start() - curRange.start() >= request.getDuration()) {
+                        timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(curRange.start(), eventRange.start(), false));                    }
```
Comment (cthamrun) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit/optional: consider moving incrementation of numNewRanges to line above and separate it from add function call for slightly clearer readability.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -14,10 +14,118 @@
...
+                    if (curRange.end() - eventRange.end() >= request.getDuration()) {
+                    timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(eventRange.end(), curRange.end(), false));
+                    }
+                    timeranges.remove(i);
+                    i += numNewRanges;
+                // case 3: slight slight overlap: either event starts before timerange or starts during and continues past timerange
```
Comment (cthamrun) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: "slight slight" typo?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (emaohn) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The variable was intended to just ensure that I don't look for extra mandatory attendees who will be attending the event and perform the same splitting computation since we'd already know the event time causes conflict for the meeting. But yes I agree the initialization is pretty redundant so I'll remove that and also rename the variable to make it more clear! Thanks for catching it.

**Comment by emaohn** [comment]
I fixed the errors that you pointed out and added some commenting for clarification although perhaps my algorithm to split and re-insert times into the time ranges is messy and incomprehensible?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 8 above)
Comment (emaohn) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:300]: Yes! I believe the test everyMandatoryAtendeesIsPrioritized() tests that case.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (ganewman42) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yeah it might not be the best practice. Admittedly I was pretty tired when I read this CL, but I found it a bit confusing. That kind of deletion sends up a red flag to a reader because it's a common mistake. Even if you've mitigated it and everything works fine, you don't want to use patterns that read as a bug to a casual observer. Think about using a different deletion pattern.

**Comment by emaohn** [comment]
I changed handleEventConflict to return a new arraylist of timeranges that I construct as I look for conflicts and resolve them by adding working ranges into the result rather than deleting and inserting into the original timerange list for better readability.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (emaohn) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Sounds good, just implemented an alternative solution.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (cthamrun) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: One good side note here is that when there are implicit/not immediately obvious implications (e.g. time ranges will be sorted), it would be good to add it as a comment to make it explicit to make things clearer.

**Review by cthamrun — APPROVED** [review]
